tests/Unit/DataStructures/Test_Mesh.cpp

```diff
@@ -0,0 +1,111 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "tests/Unit/TestingFramework.hpp"
+
+#include <array>
+
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+SPECTRE_TEST_CASE("Unit.DataStructures.Mesh", "[DataStructures][Unit]") {
+  SECTION("Uniform LGL mesh") {
+    const Mesh<1> mesh1d_lgl{3, Spectral::Basis::Legendre,
+                             Spectral::Quadrature::GaussLobatto};
+    CHECK(mesh1d_lgl.bases()[0] == Spectral::Basis::Legendre);
+    CHECK(mesh1d_lgl.quadratures()[0] == Spectral::Quadrature::GaussLobatto);
+    const Mesh<2> mesh2d_lgl{3, Spectral::Basis::Legendre,
+                             Spectral::Quadrature::GaussLobatto};
+    CHECK(mesh2d_lgl.bases()[0] == Spectral::Basis::Legendre);
+    CHECK(mesh2d_lgl.quadratures()[0] == Spectral::Quadrature::GaussLobatto);
+    CHECK(mesh2d_lgl.bases()[1] == Spectral::Basis::Legendre);
+    CHECK(mesh2d_lgl.quadratures()[1] == Spectral::Quadrature::GaussLobatto);
+    const Mesh<3> mesh3d_lgl{3, Spectral::Basis::Legendre,
+                             Spectral::Quadrature::GaussLobatto};
+    CHECK(mesh3d_lgl.bases()[0] == Spectral::Basis::Legendre);
+    CHECK(mesh3d_lgl.quadratures()[0] == Spectral::Quadrature::GaussLobatto);
+    CHECK(mesh3d_lgl.bases()[1] == Spectral::Basis::Legendre);
+    CHECK(mesh3d_lgl.quadratures()[1] == Spectral::Quadrature::GaussLobatto);
+    CHECK(mesh3d_lgl.bases()[2] == Spectral::Basis::Legendre);
+    CHECK(mesh3d_lgl.quadratures()[2] == Spectral::Quadrature::GaussLobatto);
+  }
+
+  SECTION("Explicit choices per dimension") {
+    const Mesh<1> mesh1d{
+        {{4}}, {{Spectral::Basis::Legendre}}, {{Spectral::Quadrature::Gauss}}};
+    CHECK(mesh1d[0] == 4);
+    CHECK(mesh1d.product() == 4);
```

Comment:
do we want a member function size that returns product?

Reply:
There's already a size function that returns Dim (inherited from Index, similarly to product).

Reply:
how about number_of_grid_points?